pkg/activator/net/throttler.go

```diff
@@ -0,0 +1,251 @@
+)
+
+type podIPTracker struct {
+	dest     string
+	requests int
+}
+
+type revisionThrottler struct {
+	revID                types.NamespacedName
+	containerConcurrency v1beta1.RevisionContainerConcurrencyType
+
+	// This is a breaker for the revision as a whole. try calls first pass through
+	// this breaker and are either called with clusterIPDest or go through selecting
+	// a podIPTracker and are then called.
+	breaker *queue.Breaker
+
+	// If we have a healthy clusterIPDest this is set to nil.
+	podIPTrackers []*podIPTracker
+
+	// If we dont have a healthy clusterIPDest this is set to the default, otherwise
+	// it is the l4dest for this revision's private clusterIP
+	clusterIPDest string
+
+	mux sync.RWMutex
+
+	logger *zap.SugaredLogger
+}
+
+func newRevisionThrottler(revID types.NamespacedName,
+	containerConcurrency v1beta1.RevisionContainerConcurrencyType,
+	breakerParams queue.BreakerParams) *revisionThrottler {
+	return &revisionThrottler{
+		revID:                revID,
+		containerConcurrency: containerConcurrency,
+		breaker:              queue.NewBreaker(breakerParams),
+	}
+}
+
+func (rt *revisionThrottler) try(ctx context.Context, function func(string) error) error {
+	var ret error
+
+	if !rt.breaker.Maybe(ctx, func() {
+		// See if we can get by with only a readlock
+		rt.mux.RLock()
+		if rt.clusterIPDest != "" {
+			ret = function(rt.clusterIPDest)
+			rt.mux.RUnlock()
+			return
+		}
+
+		// If we have no podIPs either then short circuit
+		if len(rt.podIPTrackers) == 0 {
+			ret = errors.New("made it through breaker but we have no clusterIP or podIPs. This should" +
+				" never happen")
+			return
+		}
+		rt.mux.RUnlock()
+
+		// We have to update podIPTrackers so take out a write lock
+		rt.mux.Lock()
+
+		// find dest with least conn
+		var leastConn *podIPTracker
+		for _, tracker := range rt.podIPTrackers {
+			if leastConn == nil || leastConn.requests > tracker.requests {
+				leastConn = tracker
+			}
+		}
+
+		// Increment requests, unlock, and run function
+		leastConn.requests++
+		rt.mux.Unlock()
+		ret = function(leastConn.dest)
+
+		// Decrement requests and return
+		rt.mux.Lock()
+		leastConn.requests--
+		rt.mux.Unlock()
+		return
+	}) {
+		return fmt.Errorf("revision %q exceeded capacity, failing request.", rt.revID)
+	}
+	return ret
+}
+
+func (rt *revisionThrottler) calculateCapacity(size, activatorCount, maxConcurrency int) int {
+	targetCapacity := int(rt.containerConcurrency) * size
+
+	if size > 0 && (rt.containerConcurrency == 0 || targetCapacity > maxConcurrency) {
+		// If cc==0, we need to pick a number, but it does not matter, since
+		// infinite breaker will dole out as many tokens as it can.
+		targetCapacity = maxConcurrency
+	} else if targetCapacity > 0 {
+		targetCapacity = minOneOrValue(targetCapacity / minOneOrValue(activatorCount))
+	}
+
+	return targetCapacity
+}
+
+// This function will never be called in parallel but try can be called in parallel to this
```

Comment:
```suggestion
// This function will never be called in parallel but `try` can be called in parallel to this
```
So do we need the lock or not?

Reply:
Updated comment explaining why